hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/StatefulServiceStateManagerImpl.java

```diff
@@ -0,0 +1,126 @@
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.hdds.scm.ha;
+
+import com.google.protobuf.ByteString;
+import org.apache.hadoop.hdds.protocol.proto.SCMRatisProtocol.RequestType;
+import org.apache.hadoop.hdds.scm.metadata.DBTransactionBuffer;
+import org.apache.hadoop.hdds.utils.db.Table;
+
+import java.io.IOException;
+import java.lang.reflect.Proxy;
+
+/**
+ * This class implements methods to save and read configurations of a
+ * stateful service from DB.
+ */
+public final class StatefulServiceStateManagerImpl
+    implements StatefulServiceStateManager {
+
+  // this table maps the service name to the configuration (ByteString)
+  private Table<String, ByteString> statefulServiceConfig;
+  private final DBTransactionBuffer transactionBuffer;
+
+  private StatefulServiceStateManagerImpl(
+      Table<String, ByteString> statefulServiceConfig,
+      DBTransactionBuffer scmDBTransactionBuffer) {
+    this.statefulServiceConfig = statefulServiceConfig;
+    this.transactionBuffer = scmDBTransactionBuffer;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void saveConfiguration(String serviceName, ByteString bytes)
+      throws IOException {
+    // do we need a write lock here?
+    // do we need the buffer here or should we directly put the key-value?
+    /*
+    Alternative: statefulServiceConfig.put(serviceName, bytes);
+     */
+    transactionBuffer.addToBuffer(statefulServiceConfig, serviceName, bytes);
+    if (transactionBuffer instanceof SCMHADBTransactionBuffer) {
+      SCMHADBTransactionBuffer buffer =
+              (SCMHADBTransactionBuffer) transactionBuffer;
+      buffer.flush();
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public ByteString readConfiguration(String serviceName) throws IOException {
+    // do we need a read lock here?
+    return statefulServiceConfig.get(serviceName);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void reinitialize(Table<String, ByteString> configs) {
+    this.statefulServiceConfig = configs;
+  }
+
+  public static Builder newBuilder() {
+    return new Builder();
+  }
+
+  /**
+   * Builder for StatefulServiceStateManager.
+   */
+  public static class Builder {
+    private Table<String, ByteString> statefulServiceConfig;
+    private DBTransactionBuffer transactionBuffer;
+    private SCMRatisServer scmRatisServer;
+
+    public Builder setStatefulServiceConfig(
+        final Table<String, ByteString> statefulServiceConfig) {
+      this.statefulServiceConfig = statefulServiceConfig;
+      return this;
+    }
+
+    public Builder setSCMDBTransactionBuffer(
+        final DBTransactionBuffer dbTransactionBuffer) {
+      this.transactionBuffer = dbTransactionBuffer;
+      return this;
+    }
+
+    public Builder setRatisServer(final SCMRatisServer ratisServer) {
+      scmRatisServer = ratisServer;
+      return this;
+    }
+
+    public StatefulServiceStateManager build() {
```

Comment:
Add
```
Preconditions.checkNotNull(statefulServiceConfig);
Preconditions.checkNotNull(transactionBuffer);
Preconditions.checkNotNull(scmRatisServer);
```

Reply:
Can we exclude the not null check for `scmRatisServer`? `SCMHAInvocationHandler` takes care of the null case by calling the `invokeLocal` method.